rfcs/20231108-gpt-quantization/README.md

```diff
@@ -0,0 +1,136 @@
+size and bandwidth requirements.
+
+The quantization details:
+- The potential weights compression data types are integer 8, 4, 3 and even 2 bits.
+- The quantization is asymmetric (both scales and zero points are used).
+- The quantization scales and zero points are multidimensional and contain groups.
+
+## OpenVino support
+
+Recently OpenVino added support for the following receipts of weights compression to support GPT-Q:
+1. int8 weights compression:
+   - Scales are per-output channels, data type is f16/f32
+   - Zero points are per-output channels, data type is u8/s8
+2. int4 weights compression [[#2]][2] [[#3]][3]:
+   - Scales are per-output and per-input channels with groups along input channels, data type is f16/f32
+   - Zero points are per-output and per-input channels with groups along input channels, data type is u8/s8
+
+# 2. Implications to oneDNN
+
+There are 3 new quantization features that can't be supported by the current oneDNN API:
+- scales and zero points data type. Currently oneDNN defines scales as f32 data and zero points as s32 data.
+- scales and zero points groups support. Currently scales and zero points are set using a mask where each
+  bit represents if for a given dimension multiple or a single scale/zero point is used.
+- Weights quantization applied to primitives with non-integer activations. This part is covered by a separate RFC: [[#4]][4]
+
+# 3. Proposal
+
+## Scales and zero points data type support
+
+### Option a (recommended)
+
+Extend scales and zero points setting API with a parameter for data type. This additional information will allow kernels to generate
+the best code for a given data type.
+
+```cpp
+// C API
+status_t dnnl_primitive_attr_set_scales(
+        primitive_attr_t *attr, int arg, ..., dnnl_data_type_t data_type);
+
+// C++ API
+void set_scales_mask(int arg, ..., data_type_t data_type);
+
+// An example
+primitive_attr attr;
+attr.set_scales(DNNL_ARG_DST, ..., dt::f16);
+...
+auto dst_scales_md = memory::desc({1}, dt::f16, tag::x);
+auto dst_scales = memory(dst_scales_md, engine);
+...
+p.execute(s, {{DNNL_ARG_ATTR_SCALES | DNNL_ARG_DST, dst_scales}, ...});
+```
+
+### Option b
+
+Keep data types implicit during scales and zero points setup, but detect the data type from the memory objects passed at execution.
+
+```cpp
+// An example
+primitive_attr attr;
+attr.set_scales(DNNL_ARG_DST, ...);
+...
+auto dst_scales_md = memory::desc({1}, dt::f16, tag::x);
+auto dst_scales = memory(dst_scales_md, engine);
+...
+p.execute(s, {{DNNL_ARG_ATTR_SCALES | DNNL_ARG_DST, dst_scales} ...});
+```
+
+## Scales and zero points groups support
+
+## Option a (recommended)
+
+Extend scales and zero points setting API with a parameter for groups.
+
+```cpp
+// C API
+status_t dnnl_primitive_attr_set_scales(primitive_attr_t *attr, int arg,
+        int mask, int ndims, const dims_t groups, ...);
+
+// C++ API
+void set_scales_mask(int arg, int mask, dims groups, ...);
+```
```

Comment:
Could you clarify the semantic of groups here? For example, we have a 1D tensor of size 1024, and the user needs to use one scale per 16 elements. Do they pass the size of groups (`groups = {16}`) or the number of groups (`groups = {1024 / 16}`) ? If the former, what does the user need to pass if there are no groups ?
If we want to make the attribute setting independent of input/output shapes, we will likely need to use a special value for this right ?

Also, what is the broadcast semantic (I guess scales[0] is used for tensor[0... groups[0]-1]) ? What do we do if the groups argument does not divide the dimension ?

Reply:
> Could you clarify the semantic of groups here? For example, we have a 1D tensor of size 1024, and the user needs to use one scale per 16 elements. Do they pass the size of groups (groups = {16}) or the number of groups (groups = {1024 / 16}) ? If the former, what does the user need to pass if there are no groups ?
If we want to make the attribute setting independent of input/output shapes, we will likely need to use a special value for this right ?

I agree, It is unclear from the description, `groups` is a group_size for each dimension.
For example, for tensor dims `[32, 64]`:
If groups are `[4, 1]` then scale dims are `[8, 64]`
So there is no a special value, for independent scales groups[i] = 1

> What do we do if the groups argument does not divide the dimension ?

oneDNN will return invalid_arguments at primitive_desc creation time

Reply:
Does this mean we will return `invalid_arguments` for DNNL_RUNTIME_DIM? Or would the check be deferred to execution in that case?

> So there is no a special value, for independent scales groups[i] = 1

I guess the question was about the case where there is a common scale for a dimension. In your example, it means that scale dims are `[8, 1]`, but groups are `[4, 64]`, in which case the group value depends on the tensor dimension. If we want to truly make groups values independant of tensor dims, we will need a special value in that case.
(We discussed that offline but I am putting it here for documentation.)

Reply:
BTW, do we really still need mask when groups are passed?

Reply:
Yeah, sorry for the confusion, I made this comment before we discussed it offline.

I updated the RFC with the latest version, please take a look.
I removed the mask.

Reply:
@mgouicem Actually having a mask would remove the need for a special group_size value:
- If i-th bit in the mask is not set, the groups[i] is ignored.
- If i-th bit in the mask is set, the group_size is taken into account and we would use groups[i]=1 for per_dim semantics or groups[i]=G for groups scales/zero points